aten/src/ATen/native/cpu/int4mm_kernel.cpp

```diff
@@ -773,10 +778,409 @@ void int4pack_mm_kernel(
   }
 }
 
+/**
+ * The Int4 quantized weights must be represented as a uint8 tensor
+ * For matrix multiplication with a weight shape of (N x K)
+ * the shape of the 4-bit quantized weights is [N, K/groupsize, groupsize/2].
+ *
+ * For KleidiAI weight packing, the scales, biases, and Int4 quantized
+ * weights are packed into a single `packed_weights` structure, optimized for
+ * Arm instructions.
+ *
+ * In the fallback reference kernel, no special packing is required for
+ * Int4 quantized weights.
+ */
+void dyn_quant_pack_4bit_weight_kernel(
+    Tensor& packed_weights,
+    const Tensor& weights,
+    const Tensor& scales_zeros,
+    const Tensor& bias,
+    const int64_t N,
+    const int64_t K,
+    const int64_t block_size) {
+#if AT_KLEIDIAI_ENABLED()
+  const int64_t weight_packed_size =
+      kleidiai::kai_pack_rhs_int4_size(N, K, block_size);
+  packed_weights.resize_({weight_packed_size});
+  kleidiai::kai_pack_int4_rhs(
+      packed_weights, weights, scales_zeros, bias, N, K, block_size);
+#else
+  TORCH_CHECK(
+      packed_weights.sizes() == weights.sizes(),
+      __func__,
+      " : Packed weights shape should be same as weights");
+  packed_weights.copy_(weights);
```

Comment:
We do not concat multiple tensors to make it explicity for the user via signature. We want to keep the signature more elaboraitve so that it can accommodate other contributors requirements